### DISCUSSION TITLE
PlanetScale is currently not suitable for serverless edge applications

### DISCUSSION THREAD

**Discussion by stewarthsoj** [planetscale/discussion#107] [Feature Requests] [answered]
I've been playing around with PlanetScale for a few weeks now and it's given me some time to think about how it could be made more suitable for applications hosted on edge computing platforms i.e. Cloudflare Workers and Deno Deploy. The problem at the moment is these environments don't support Node.js-dependent libraries including `mysql`, which is basically a standard library for connecting a Node.js app to PlanetScale. Databases such as FaunaDB, Dynamo, and Firebase provide a HTTP interface, which makes them well-suited for these edge compute environments.

But, there is also another consideration - database centralisation. At the moment, PlanetScale databases can only be deployed to one region. This defeats the purpose of attempting to connect an application on the edge to a PlanetScale database. As far as I'm aware, PlanetScale databases can't be deployed in multiple regions.

The following has to be supported before PlanetScale is a suitable option for serverless edge applications:

- Support for a HTTP interface
- Deploy a PlanetScale database to multiple regions (writes and reads should be distributed)

Knowing this, is there anyone from PlanetScale that could shed some light on the above and perhaps provide an update as to whether they'll be supporting these features in the near future?

**Comment by nickvanw** [comment, marked as answer]
Hey @JoshStwrt! I really appreciate the thoughtful post.

When connecting connecting to PlanetScale, any MySQL driver should work without issue - this means that on Deno, for example, the [MySQL Driver](https://github.com/denodrivers/mysql) could work. It looks like this might not support TLS, which would be a blocker to connecting, as PlanetScale requires it for security purposes. This looks to be a [known problem](https://github.com/denodrivers/mysql/issues/118), and I would love to see if we can help them make any progress. 

For other platforms that don't allow any outgoing TCP connections except for HTTP, though, you're right that it makes using PlanetScale difficult without an intermediary API. We are looking at figuring out how to unblock this in platforms like Cloudflare Workers, but don't have any imminent plans to release a generic database HTTP API for the PlanetScale platform itself.  

Database centralization is definitely a bigger issue, and it's one that we are going to solve long-term with [Vitess' Sharding](https://vitess.io/docs/reference/features/sharding/) support, which allows for a database to be split up into different cells that can live around this world. There is no 'free lunch' when it comes to the speed of light, though - every solution has a performance/throughput/scaling tradeoff, and we're working hard to find the right one for our platform. 

By creating geographically-spread sharding setup, a developer can choose where they store their data, and keep it close to the user, so that it can be looked up quickly. Stay tuned for more information in this space, as we're going to be getting incrementally closer and closer to this as we develop the platform more.

Let me know if you have any other questions, and I hope that we can pave a path for you to make PlanetScale useful for you!

**Reply by fergusmeiklejohn** [reply]
Does Prisma solve the problem of connecting to PlanetScale from Cloudflare workers?

The edge database problem is very difficult to solve, but it is something that we're worried about too. We plan to deploy to CloudFlare to get all the benefits of the edge, but need a database solution that works well in that environment.

**Reply by 2color** [reply]
Hey @fergusmeiklejohn,

Using Prisma with the Prisma Data Proxy will allow you to use PlanetScale from Cloudflare workers. 

You can find the guide here: https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-cloudflare-workers

Note that this is currently Early Access so I would be wary of adopting this for production workloads at this point.

**Reply by fergusmeiklejohn** [reply]
Thanks :-) I was just watching this from the Prisma serverless conf: https://www.youtube.com/watch?v=kOTTQsOBqdY
Seems like CloudFlare and/or Prisma also have solutions in the pipeline for the problem of accessing our data at the edge

**Reply by paambaati** [reply]
@nickvanw Is there a ticket where can follow progress? Distributed (or geo-partitioned, or data replicated at the edge) is something we're super-interested in, and I'd love to learn more about PlanetScale's plans for this.

**Reply by jgreet** [reply]
We don't have any announcements to make yet, but a good place to watch for major new features is our change log: 
https://planetscale.com/changelog

We'll also post an update to this thread when we have something relevant to share.

**Reply by tbarn** [reply]
Just in case anyone missed it, we released the serverless driver a few weeks ago: https://planetscale.com/blog/introducing-the-planetscale-serverless-driver-for-javascript

This new driver and infrastructure update brings you:
- The ability to store and query data in PlanetScale from environments such as Cloudflare Workers, Vercel Edge Functions, and Netlify Edge Functions.
- Infrastructure improvements that enable global routing and improve connection reliability and performance.
